source/particle/world.cc

```diff
@@ -0,0 +1,889 @@
+
+      fe_value.reinit (cell);
+      fe_value.get_function_values (this->get_solution(),
+                                    values);
+      fe_value.get_function_gradients (this->get_solution(),
+                                       gradients);
+
+      it = begin_particle;
+      for (unsigned int i = 0; it!=end_particle; ++it,++i)
+        {
+          property_manager->update_one_particle(it->second,
+                                                values[i],
+                                                gradients[i]);
+        }
+    }
+
+    template <int dim>
+    void
+    World<dim>::local_advect_particles(const typename DoFHandler<dim>::active_cell_iterator &cell,
+                                       const typename std::multimap<types::LevelInd, Particle<dim> >::iterator &begin_particle,
+                                       const typename std::multimap<types::LevelInd, Particle<dim> >::iterator &end_particle)
+    {
+      const unsigned int particles_in_cell = std::distance(begin_particle,end_particle);
+
+      std::vector<Tensor<1,dim> >  result(particles_in_cell);
+      std::vector<Tensor<1,dim> >  old_result(particles_in_cell);
+      std::vector<Point<dim> >     particle_points(particles_in_cell);
+
+      typename std::multimap<types::LevelInd, Particle<dim> >::iterator it = begin_particle;
+      for (unsigned int i = 0; it!=end_particle; ++it,++i)
+        {
+          const Point<dim> position = it->second.get_location();
+          particle_points[i] = this->get_mapping().transform_real_to_unit_cell(cell, position);
+        }
+
+      const std::vector< double > weights(particles_in_cell, 1./((double) particles_in_cell));
+      const Quadrature<dim> quadrature_formula(particle_points, weights);
+      FEValues<dim> fe_value (this->get_mapping(),
+                              this->get_fe(),
+                              quadrature_formula,
+                              update_values);
+
+      fe_value.reinit (cell);
+      fe_value[this->introspection().extractors.velocities].get_function_values (this->get_solution(),
+                                                                                 result);
+      fe_value[this->introspection().extractors.velocities].get_function_values (this->get_old_solution(),
+                                                                                 old_result);
+
+      integrator->local_integrate_step(begin_particle,
+                                       end_particle,
+                                       old_result,
+                                       result,
+                                       this->get_old_timestep());
+    }
+
+    template <int dim>
+    void
+    World<dim>::generate_particles()
+    {
+      particles = generator->generate_particles();
+    }
+
+    template <int dim>
+    void
+    World<dim>::initialize_particles()
+    {
+      // Loop over all cells and initialize the particles cell-wise
```

Comment:
this function is a pretty good candidate for the WorkStream interface. No need to do this now, but maybe leave a TODO about this? Same with the update function.